File: two-pointers/0287-find-the-duplicate-number.py
class Solution(object):
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        slow = 0
        fast = 0
        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            print(11, slow, fast)

            if slow==fast:
                break
        print(slow, fast)


        slow2=0
        while True:
            slow = nums[slow]
            slow2 = nums[slow2]
            if slow==slow2:
                print(slow)
                break


nums = [1,3,4,2,2]
Solution().findDuplicate(nums)

File: two-pointers/0287-find-the-duplicate-number.go
package main

import "fmt"

func findDuplicate(nums []int) int {
	slow, fast := 0, 0
	for true {
		slow = nums[slow]
		fast = nums[nums[fast]]
		if slow == fast {
			break
		}

	}
	slow2 := 0
	for true {
		slow = nums[slow]
		slow2 = nums[slow2]
		if slow == slow2 {
			return slow
		}

	}
	fmt.Println(slow)
	return -1

}
func main() {
	findDuplicate([]int{1, 3, 4, 2, 2})

}

// Input: nums = [1,3,4,2,2]
// Output: 2
